core/pkg/artifacts/file_cache.go

```diff
@@ -0,0 +1,163 @@
+package artifacts
+
+import (
+	"crypto/md5"
+	"encoding/base64"
+	"encoding/hex"
+	"errors"
+	"io"
+	"log/slog"
+	"os"
+	"path/filepath"
+
+	atomicfile "github.com/natefinch/atomic"
+	"github.com/wandb/wandb/core/pkg/utils"
+)
+
+type FileCache struct {
+	root string
+}
+
+func NewFileCache() *FileCache {
+	dir, found := os.LookupEnv("WANDB_CACHE_DIR")
+	if !found {
+		userCacheDir, err := os.UserCacheDir()
+		if err != nil {
+			return nil
```

Comment:
returning `nil` from a `New..` function is surprising and could be easy to miss. If it returned an error instead, it would be clear at the callsite that the result could be nil.